### PR TITLE
KIALI-2442 Map kind and apiVersion attributes

### PR DESCRIPTION
**Pull request by lucasponce**
It adds mapping for "kind" and "apiVersion" objects from the backend.
Note that these fields are not returned from the k8s API (those are empty), so it should be provided by kiali.

** Issue reference **

https://issues.jboss.org/browse/KIALI-2442

### CODE REVIEW

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#624](https://paste.fedoraproject.org/paste/aAiqHSLOYWwcH7~adIfmVA)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1335]](https://paste.fedoraproject.org/paste/LL~y-QXlTl3UQaXUGeDPfg)

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#625](https://paste.fedoraproject.org/paste/r8ZtFHGkiwO~AMAqQNC19A)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1336]](https://paste.fedoraproject.org/paste/Xc78OnbtPd6hx97DFrLxeQ)

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#626](https://paste.fedoraproject.org/paste/kSJtQ1J6i2XT7qWhU74WCA)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1337]](https://paste.fedoraproject.org/paste/12N3jp9emQ-4CWM~5Ta0lg)

File: kubernetes/istio_details_service.go [hunk 1]
```diff
@@ -99,15 +110,25 @@ func (in *IstioClient) UpdateIstioObject(api, namespace, resourceType, name, jso
 	log.Debugf("UpdateIstioObject input: %s / %s / %s / %s", api, namespace, resourceType, name)
 	var result runtime.Object
 	var err error
+
+	typeMeta := meta_v1.TypeMeta{
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply kubernetes/istio_details_service.go]: Yes, I forgot the assignment here.

File: kubernetes/istio_details_service.go [hunk 2]
```diff
@@ -75,6 +85,7 @@ func (in *IstioClient) CreateIstioObject(api, namespace, resourceType, json stri
 	if !ok {
 		return nil, fmt.Errorf("%s/%s doesn't return an IstioObject object", namespace, resourceType)
 	}
+	istioObject.SetTypeMeta(typeMeta)
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply kubernetes/istio_details_service.go]: No, here it is needed.
As the CreateIstioObject returns an IstioObject.
typeMeta is not a value here, it's a composition that simplifies to bring the "kind" and "apiVersion" attributes.

File: kubernetes/istio_details_service.go [hunk 3]
```diff
@@ -151,12 +177,17 @@ func (in *IstioClient) GetVirtualService(namespace string, virtualservice string
...
+		Kind:       PluralType[virtualServices],
+		APIVersion: ApiNetworkingVersion,
+	}
 	virtualService, ok := result.(*GenericIstioObject)
 	if !ok {
 		return nil, fmt.Errorf("%s/%s doesn't return a VirtualService object", namespace, virtualservice)
 	}
-	return virtualService.DeepCopyIstioObject(), nil
+	vs := virtualService.DeepCopyIstioObject()
+	vs.SetTypeMeta(typeMeta)
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply kubernetes/istio_details_service.go]: No, it is not, TypeMeta are empty for some objects due a limitation in the k8s.
This is why we need to assign it after the call.

File: kubernetes/istio_rules_service.go [hunk 4]
```diff
@@ -11,14 +14,19 @@ func (in *IstioClient) GetIstioRules(namespace string) ([]IstioObject, error) {
...
 	ruleList, ok := result.(*GenericIstioObjectList)
 	if !ok {
 		return nil, fmt.Errorf("%s doesn't return a rules list", namespace)
 	}
-
 	istioRules := make([]IstioObject, 0)
 	for _, rule := range ruleList.Items {
-		istioRules = append(istioRules, rule.DeepCopyIstioObject())
+		r := rule.DeepCopyIstioObject()
+		r.SetTypeMeta(typeMeta)
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply kubernetes/istio_rules_service.go]: No, I guess you missed something.
```
istioRule.TypeMeta = rule.GetTypeMeta()
```
models.CastIstioRule parses TypeMeta to bring "kind" and "apiVersion" to the model.

**Comment by lucasponce** [comment]
I have a pending commit to address one of your comments.
I'll send it shortly, after I test everything locally.

**Comment by lucasponce** [comment]
@xeviknal fixed, let me know if there is anything else related this work. Thx.

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#633](https://paste.fedoraproject.org/paste/ivkV57hA5a5im3BX7F4hzQ)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1355]](https://paste.fedoraproject.org/paste/JW7j6pcYLo29asvO-4tckw)

**Pull request merged**